File: Capacity To Ship Packages Within D Days.java
class Solution {
    public static boolean isPossible(int[] weights, int days, int mid){
        int dayCount = 1;
        int weightSum = 0;

        for(int i=0; i<weights.length; i++){
            if(weights[i] + weightSum <= mid){
                weightSum += weights[i];
            }
            else{
                dayCount++;
                weightSum = 0;

                if(weights[i] > mid || dayCount > days){
                    return false;
                }

                weightSum += weights[i];
            }
        }

        return true;
    }

    public int shipWithinDays(int[] weights, int days) {
        int sum = 0;
        for(int i=0; i<weights.length; i++){
            sum += weights[i];
        }

        int s = 0;
        int e = sum;
        int mid = s + (e-s)/2;
        int ans = 0;

        while(s<=e){
            if(isPossible(weights,days,mid)){
                ans = mid;
                e = mid-1;
            }
            else{
                s = mid+1;
            }
            mid = s + (e-s)/2;
        }

        return ans;
    }
}

File: Capacity To Ship Packages Within D Days.cpp
class Solution {
public:
    bool check(int mid, vector<int> w, int days, int n){
        int cnt = 0, temp = 0;
        for(int i=0; i<n; i++){
            temp = temp + w[i];
            if(temp > mid){
                cnt++;
                temp = w[i];
            }
        }
        if(cnt >= days){
            return true;
        }else{
            return false;
        }
    }
    int shipWithinDays(vector<int>& w, int days) {
        int n = w.size();
        int low=0,high=0;
        for(int i=0; i<n; i++){
            low = max(low,w[i]);
            high = high + w[i];
        }
        
        int mid,ans;
        while(low <= high){
            mid = low + (high-low)/2;
            if(!check(mid,w,days,n)){
                ans = mid;
                high = mid-1;
            }else{
                low = mid+1;
            }
        }
        return ans;
    }
};
